client/my-sites/patterns/components/pattern-gallery/client.tsx

```diff
@@ -14,12 +15,68 @@ import './style.scss';
 
 const LOGGED_OUT_USERS_CAN_COPY_COUNT = 3;
 
+type MasonryGalleryProps = PropsWithChildren< {
+	className?: string;
+	enableMasonry: boolean;
+} >;
+
+// Simulates a Masonry layout by applying negative `margin-top` on every item that doesn't sit in
+// the first row
+function MasonryGallery( { children, className, enableMasonry }: MasonryGalleryProps ) {
+	const ref = useRef< HTMLDivElement >( null );
+
+	useLayoutEffect( () => {
+		if ( ! ref.current || ! enableMasonry ) {
+			return;
+		}
+
+		const element = ref.current;
+
+		function calculateLayout() {
+			const columnCount = getComputedStyle( element ).gridTemplateColumns.split( ' ' ).length;
+
+			if ( columnCount === 1 ) {
+				return;
+			}
+
+			const items = [ ...element.querySelectorAll< HTMLElement >( '.pattern-preview' ) ];
+
+			// We calculate the difference between the top coordinates of each `.pattern-preview`
+			// with the bottom coordinates of the first `.pattern-preview` in the same column. This
+			// value is then used to set a negative `margin-top`, simulating a Masonry layout
+			items.slice( columnCount ).forEach( ( item, i ) => {
+				item.style.marginTop = '0';
```

Comment:
Instead of setting `marginTop` twice, can we just remove the current `marginTop` from the current value?

Reply:
I take it you mean that we could subtract `marginTop` from `getBoundingClientRect().bottom` or something like that? We actually need to reset `marginTop` to handle breakpoints where the number of grid columns changes, so I would keep this as is.

Reply:
I took a stab at this after all and came up with a solution that works quite well: we always reset the margin for the items in the first row but not for subsequent ones. For those, we just add the existing top margin to the calculation, which saves the browser some redrawing. Thanks for the suggestion, @tyxla 👍